File: src/problems/4/s0.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        return recur(nums1, nums2)


def recur(nums1, nums2):
    if len(nums1) == 1 and len(nums2) == 1:
        return (nums1[0] + nums2[0]) / 2

    mid1 = len(nums1) // 2
    mod1 = len(nums1) % 2
    med1 = (nums1[mid1] + nums1[mid1 - 1]) / 2 if mod1 == 0 else nums1[mid1]

    mid2 = len(nums2) // 2
    mod2 = len(nums2) % 2
    med2 = (nums2[mid2] + nums2[mid2 - 1]) / 2 if mod2 == 0 else nums2[mid2]

    if med1 == med2:
        return med1
    elif med1 > med2:
        return recur(nums1[: mid1 + mod1], nums2[mid2:])
    elif med1 < med2:
        return recur(nums1[mid1:], nums2[: mid2 + mod2])


def main():
    solution = Solution()
    # nums1 = [1, 3]
    # nums2 = [2]
    # print(solution.findMedianSortedArrays(nums1, nums2))

    nums1 = [1, 2]
    nums2 = [3, 4]
    print(solution.findMedianSortedArrays(nums2, nums1))


if __name__ == "__main__":
    main()

File: src/problems/4/s0.rs
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        recur(nums1, nums2)
    }
}

fn recur(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
    if nums1.len() == 1 && nums2.len() == 1 {
        (nums1[0] + nums2[0]) as f64 / 2
    }

    let mid1 = nums1.len() / 2;
    let mod1 = nums1.len() % 2;
    let med1 = if mod1 == 0 {
        (nums1[mid1] + nums1[mid1 - 1]) as f64 / 2
    } else {
        nums1[mid1] as f64
    };

    let mid2 = nums2.len() / 2;
    let mod2 = nums2.len() % 2;
    let med2 = if mod2 == 0 {
        (nums2[mid2] + nums2[mid2 - 1]) as f64 / 2
    } else {
        nums2[mid2] as f64
    };

    if med1 == med2 {
        med1
    } else if med1 > med2 {
        recur(&nums1[..mid1 + mod1].to_vec(), &nums2[mid2..].to_vec())
    } else {
        recur(&nums1[mid1..].to_vec(), &nums2[..mid2 + mod2].to_vec())
    }
}
